File: codewars/8-kyu/wilson-primes/solution.py
# https://www.codewars.com/kata/55dc4520094bbaf50e0000cb

# solution 1: 547ms
def am_i_wilson(n):
    if (n < 2) or (n % 2 == 0) or not all(n % i for i in range(3, int(n ** 0.5) + 1, 2)):
        return False
    
    return (mod_factorial(n - 1) + 1) % (n ** 2) == 0

def mod_factorial(n, m):
    result = 1
    for i in range(1, n+1):
        result = (result * i) % m 
    
    return result 

# solution 2:
# the only known wilson primes are 5, 13, 563
def am_i_wilson(n):
    return n in (5, 13, 563)

File: codewars/8-kyu/wilson-primes/solution.js
// https://www.codewars.com/kata/55dc4520094bbaf50e0000cb

// solution 1: 812ms
function amIWilson(p) {
    let m = p * p;
    return (mod_factorial(p-1, m) + 1) % m === 0;
}

function mod_factorial(n, m) {
    let result = 1;
    for (i = 1; i <= n; i++) {
        result = (result * i) % m;
    }
    return result;
}

// solution 2:
// the only known Wilson primes are 5, 13, 563
function amIWilson(p) {
    return [5, 13, 563].indexOf(p) > -1;
}